From 68053f7715301b9b75f748e163e42a586eca191c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Roger=20Pau=20Monn=C3=A9?= <roger.pau@citrix.com>
Date: Tue, 30 Jun 2026 14:56:44 +0200
Subject: riscv/irq: define a per-arch irq_to_desc()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

RISCV declares irq_desc as a static array, opposed to x86 that uses a
pointer allocated at boot time.  This creates issues when attempting to add
an extern declaration for irq_desc, as asm/irq.h is included by xen/irq.h where
the definition of struct irq_desc resides, and an empty forward declaration
doesn't make the compiler happy because it doesn't know the type
data-storage.

Signed-off-by: Roger Pau Monné <roger.pau@citrix.com>
Reviewed-by: Oleksii Kurochko <oleksii.kurochko@gmail.com>
master commit: b9b4c6a7abdc6f8ef83592eb56d3bf9523a7536d
master date: 2026-06-29 17:14:47 +0200

diff --git a/xen/arch/riscv/include/asm/irq.h b/xen/arch/riscv/include/asm/irq.h
index f633636dc308..62648bdc4252 100644
--- a/xen/arch/riscv/include/asm/irq.h
+++ b/xen/arch/riscv/include/asm/irq.h
@@ -35,6 +35,9 @@ struct arch_irq_desc {
     unsigned int type;
 };
 
+struct irq_desc *irq_to_desc(unsigned int irq);
+#define irq_to_desc irq_to_desc
+
 struct cpu_user_regs;
 struct dt_device_node;
 
diff --git a/xen/arch/riscv/irq.c b/xen/arch/riscv/irq.c
index 25d329500212..b5066fc3e981 100644
--- a/xen/arch/riscv/irq.c
+++ b/xen/arch/riscv/irq.c
@@ -19,6 +19,12 @@
 
 static irq_desc_t irq_desc[NR_IRQS];
 
+struct irq_desc *irq_to_desc(unsigned int irq)
+{
+    ASSERT(irq < ARRAY_SIZE(irq_desc));
+    return &irq_desc[irq];
+}
+
 static bool irq_validate_new_type(unsigned int curr, unsigned int new)
 {
     return curr == IRQ_TYPE_INVALID || curr == new;
